### ISSUE TITLE
morty proxied link not working correctly

### ISSUE THREAD

**Issue by Kweck** [searx/searx-docker#18]
Hi,

if I open the proxied link of a search result I just see "page not found" within my searx page.
The generated link looks like this:
https://<my.url.here>/morty/?mortyurl=https%3A%2F%2Fde.wikipedia.org%2Fwiki%2FProxy&mortyhash=c5e85c644c08e1be9bf580d8deb5c48ad35f0e8cb5a2a19f26e6079cb215b795

**Comment by dalf** [comment]
Can you post here a screenshot of the "page not found" page ?

```http://localhost:8888/morty/?mortyurl=https%3A%2F%2Fde.wikipedia.org%2Fwiki%2FProxy&mortyhash=c5e85c644c08e1be9bf580d8deb5c48ad35f0e8cb5a2a19f26e6079cb215b795```
works as expected. 

Same on my instance and searx.me

**Comment by Kweck** [comment]
Sorry for the late response.
Here it is.


![Anmerkung 2019-12-14 182349](https://user-images.githubusercontent.com/8403406/70852207-205fbd00-1e9f-11ea-9a9f-5f9017b85a83.png)

**Comment by Kweck** [comment]
I've just seen that your instance changes the port to 3000
That's not the case on mine. But if I add the port number to the url the page isn't reachable.

**Comment by aayio** [comment]
@dalf thank you for producing these tools.

I am encountering a similar issue. When I follow a proxied link, it is a blank white page (but my browser seems to 'load' it fine - it doesn't hang or display any errors).

Everything is configured as per the instructions and I have verified that the key in settings.yml matches the key I specified in .env

Any help is greatly appreciated.

**Comment by netstx** [comment]
@dalf I am having the same issue as @mauvity described, installed everything according to instructions. Here is a snippet from /var/log/syslog:

`Jul 29 15:06:44 searx sh[14686]: #033[36mcaddy            |#033[0m {"level":"error","ts":1596053204.6043708,"logger":"http.log.error.log0","msg":"read tcp 127.0.0.1:45058->127.0.0.1:3000: read: connection reset by peer","request":{"method":"GET","uri":"/morty/?mortyurl=https%3A%2F%2Fhub.docker.com%2Fr%2Fwonderfall%2Fsearx%2F&mortyhash=97556d20753968385a3276f9fbeb7272d96d1a50119547f7ebba9fc372b07811","proto":"HTTP/2.0","remote_addr":"143.45.66.78:11977","host":"searx.thinkpad.io","headers":{"User-Agent":["Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:78.0) Gecko/20100101 Firefox/78.0"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"],"Accept-Language":["en-US,en;q=0.5"],"Accept-Encoding":["gzip, deflate, br"],"Dnt":["1"],"Cookie":["categories=general; language=en-US; locale=en; autocomplete=google; image_proxy=1; method=POST; safesearch=0; theme=oscar; results_on_new_tab=1; doi_resolver=oadoi.org; oscar-style=logicodev; disabled_engines=; enabled_engines=; disabled_plugins=; enabled_plugins=; tokens="],"Upgrade-Insecure-Requests":["1"],"Te":["trailers"]},"tls":{"resumed":true,"version":772,"ciphersuite":4867,"proto":"h2","proto_mutual":true,"server_name":"searx.thinkpad.io"}},"duration":0.003542593,"status":502,"err_id":"9iamhw82k","err_trace":"reverseproxy.(*Handler).ServeHTTP (reverseproxy.go:411)"}
`

My searx/settings.yml has this portion commented out at the beginning of the file:

`result_proxy:`
`url : http://127.0.0.1:3000/`
`key : !!binary "your_morty_proxy_key"`

and towards the end of the file, the morty configuration is:

`result_proxy:`
`url : https://searx.thinkpad.io/morty/`
`key : !!binary "my_actual_key_here"`

Any ideas on what could be causing the problem? 
Any help is much appreciated.

**Comment by dalf** [comment]
@mauvity I need some logs from the different docker images: ```docker logs caddy```. Be careful to read the log and see if there is any personnal data (it should be the case, just double check).

@raffieyeah the important message is ```read tcp 127.0.0.1:45058->127.0.0.1:3000: read: connection reset by peer```.
what does the morty logs say ? But you have to remove these lines before and restart morty (or searx-docker.service) : https://github.com/searx/searx-docker/blob/master/docker-compose.yaml#L72-L73

**Comment by netstx** [comment]
@dalf, i removed those lines from docker-compose.yaml and restarted the stack with systemd. I didn't find any morty logs in syslog, so I tried "docker logs morty" and here is what I got:

`2020/07/29 20:46:22 listening on 127.0.0.1:3000`

I also tried "nmap -sS -sV -O -p 3000 localhost" and the port is open.

**Comment by dalf** [comment]
* Can you make some request to morty, and see the log again ?
* ```curl http://127.0.0.1:3000/?mortyurl=https%3A%2F%2Fhub.docker.com%2Fr%2Fwonderfall%2Fsearx%2F&mortyhash=97556d20753968385a3276f9fbeb7272d96d1a50119547f7ebba9fc372b07811``` do you get an error ?

**Comment by netstx** [comment]
I get "curl: (52) Empty reply from server" returned from the command. Now, my VM's hostname is "searx" (Debian Buster), I wonder if this would cause a conflict somehow?

**Comment by dalf** [comment]
> Now, my VM's hostname is "searx" (Debian Buster), I wonder if this would cause a conflict somehow?

I don't think so.

> "curl: (52) Empty reply from server

Can you try again with ````-vvvv````: 
```curl -vvvv .....```
?

**Comment by netstx** [comment]
Here is the output:
```
* Expire in 0 ms for 6 (transfer 0x562ea8d1ebb0)
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x562ea8d1ebb0)
* Connected to 127.0.0.1 (127.0.0.1) port 3000 (#0)
> GET /?mortyurl=https%3A%2F%2Fgithub.com%2F HTTP/1.1
> Host: 127.0.0.1:3000
> User-Agent: curl/7.64.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact
curl: (52) Empty reply from server
```
I don't see anything in the logs (/var/log/syslog) and "docker logs morty" still only shows that one line stating it is listening on localhost:3000.

**Comment by netstx** [comment]
Here is a request to morty via Caddy, using the URL from the Searx search result:

```
*   Trying 192.154.3.133...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x558793094bb0)
* Connected to searx.thinkpad.io (192.154.3.133) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_CHACHA20_POLY1305_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=searx.thinkpad.io
*  start date: Jul 29 18:49:13 2020 GMT
*  expire date: Oct 27 18:49:13 2020 GMT
*  subjectAltName: host "searx.thinkpad.io" matched cert's "searx.thinkpad.io"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x558793094bb0)
> GET /morty/?mortyurl=https%3A%2F%2Fgithub.com%2F HTTP/2
> Host: searx.thinkpad.io
> User-Agent: curl/7.64.0
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
< HTTP/2 502
< cache-control: no-cache, no-store
< content-security-policy: default-src 'none'; style-src 'self' 'unsafe-inline'; form-action 'self'; frame-ancestors 'self'; base-uri 'self'; img-src 'self' data:; font-src 'self'; frame-src 'self'
< pragma: no-cache
< server: Caddy
< content-length: 0
< date: Wed, 29 Jul 2020 21:07:43 GMT
<
* Connection #0 to host searx.thinkpad.io left intact
```

**Comment by dalf** [comment]
It seems there is something on the way that blocks the request sent to http://localhost:3000.

**Comment by netstx** [comment]
> 
> 
> It seems there is something on the way that blocks the request sent to http://localhost:3000.

Yeah it's definitely odd... Here is netstat output, not sure if this helps:

```
netstat -tupan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:4040          0.0.0.0:*               LISTEN      1858/docker-proxy
tcp        0      0 127.0.0.1:4041          0.0.0.0:*               LISTEN      1827/docker-proxy
tcp        0      0 127.0.0.1:3000          0.0.0.0:*               LISTEN      1790/docker-proxy
tcp6       0      0 :::80                   :::*                    LISTEN      1815/caddy
tcp6       0      0 :::443                  :::*                    LISTEN      1815/caddy

```

So looks like all ports are open as expected. I don't have IPv6 enabled on this VM, so I removed "-ipv6" from the morty command in the docker-compose.yml and that didn't help either.

**Comment by aayio** [comment]
> logs from the different docker images: `docker logs caddy`

Hi @dalf , again thank you for your work and being so responsive. My error message is slightly different(?) to @raffieyeah 's. 

`{"level":"error","ts":1596090905.4998772,"logger":"http.log.error.log0","msg":"EOF","request":{"method":"GET","uri":"/morty/?mortyurl=https%3A%2F%2Fwww.qld.gov.au%2Fhealth%2Fconditions%2Fhealth-alerts%2Fcoronavirus-covid-19%2Fstay-informed%2Ftesting-and-fever-clinics&mortyhash=97e6cdc47cd4e09daf0bc778f29f4f49c8f0569d02217fda677e05dc74611630","proto":"HTTP/2.0","remote_addr":"<redacted>","host":"<searx_hostname>","headers":{"User-Agent":<redacted>,"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"],"Accept-Language":["en-au"],"Accept-Encoding":["gzip, deflate, br"]},"tls":{"resumed":false,"version":772,"ciphersuite":4865,"proto":"h2","proto_mutual":true,"server_name":"<searx_hostname>"}},"duration":0.001355151,"status":502,"err_id":"n46rw8314","err_trace":"reverseproxy.(*Handler).ServeHTTP (reverseproxy.go:411)"}`

Further logs:

```
$ sudo docker logs morty
2020/07/30 07:33:19 listening on 127.0.0.1:3000
```

```
$ curl -vvvv http://127.0.0.1:3000/?mortyurl=<redacted>
[1] 6204
* Expire in 0 ms for 6 (transfer 0x55ae3337ef50)
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x55ae3337ef50)
* Connected to 127.0.0.1 (127.0.0.1) port 3000 (#0)
> GET /?mortyurl=<redacted> HTTP/1.1
> Host: 127.0.0.1:3000
> User-Agent: curl/7.64.0
> Accept: */*
> 
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact
curl: (52) Empty reply from server
```

**Comment by aayio** [comment]
I managed to fix this, which I find absurd given how much of a noob I am.

**tl;dr**

1. Remove `-listen 0.0.0.0:3000` from `command:`
2. Add `MORTY_ADDRESS=0.0.0.0:3000` to `environment:`

So that `docker-compose.yaml` for morty looks like:

```
  morty:
    container_name: morty
    image: dalf/morty
    restart: always
    ports:
      - "127.0.0.1:3000:3000"
    networks:
      - searx
    command: -timeout 6 -ipv6
    environment:
      - MORTY_KEY=${MORTY_KEY}
      - MORTY_ADDRESS=0.0.0.0:3000
    logging:
      driver: none
    read_only: true
    cap_drop:
      - ALL
```

How I figured this out:

- Googled `"docker" "empty reply"` and found [this useful thread](https://github.com/moby/moby/issues/2522)
- Hypothesised that the error we are having with morty was due to morty not getting the instruction to listen at `0.0.0.0:3000` (i.e. on all networks) and instead just listening on `127.0.0.1:3000` (i.e. localhost _within its container only_)
- Looked at the [morty documentation](https://github.com/asciimoo/morty), saw the environment variable `MORTY_ADDRESS` which defaults to `127.0.0.1:3000`, which is fine when morty is installed on the base OS, but in the case of a Docker setup needs to be changed to `0.0.0.0:3000`

Is this a reasonable solution to the problem?

**Comment by netstx** [comment]
@mauvity, that makes sense since the other containers in the compose file also use 0.0.0.0. I tested your change and it fixed the issue on my instance, nice!

If @dalf agrees with this fix, maybe you could submit a PR?

EDIT: it's interesting that morty was working for other folks using the same docker compose file supposedly?

**Comment by dalf** [comment]
Thank you for the details.

It is a morty issue for different things:
* the docker image should run out of the box: it should listen on 0.0.0.0 by default inside the docker image.
* the ```-listen``` parameter doesn't seems to work (even outside docker).

**Comment by netstx** [comment]
> 
> 
> Thank you for the details.
> 
> It is a morty issue for different things:
> 
>     * the docker image should run out of the box: it should listen on 0.0.0.0 by default inside the docker image.
> 
>     * the `-listen` parameter doesn't seems to work (even outside docker).

Did this happen to you last time you used this docker-compose file, could it be certain setups work fine?

**Comment by dalf** [comment]
> Did this happen to you last time you used this docker-compose file, could it be certain setups work fine?

Perhaps related to this commit https://github.com/searx/searx-docker/commit/bdf5619765ac9457eb0bb4107dba2db10eddd724

**Comment by netstx** [comment]
> 
> 
> > Did this happen to you last time you used this docker-compose file, could it be certain setups work fine?
> 
> Perhaps related to this commit [bdf5619](https://github.com/searx/searx-docker/commit/bdf5619765ac9457eb0bb4107dba2db10eddd724)

I see how it ties to #38 - it makes sense. It's interesting that disabling the docker userland proxy didn't fix the issue, but the change in the docker-compose to use a morty variable worked. I submitted a PR, but I don't think that might be the greatest fix (although it does work).

**Comment by dalf** [comment]
Close by #51
